src/main/kotlin/cacophony/codegen/instructions/matching/InstructionMatcher.kt

```diff
@@ -11,3 +12,120 @@ interface InstructionMatcher {
 
     fun findMatchesForCondition(node: CFGNode, destinationLabel: BlockLabel, jumpIf: Boolean = true): Set<Match>
 }
+
+class InstructionMatcherImpl(
+    private val valuePatterns: List<ValuePattern>,
+    private val sideEffectPatterns: List<SideEffectPattern>,
+    private val conditionPatterns: List<ConditionPattern>,
+) : InstructionMatcher {
+    override fun findMatchesForValue(node: CFGNode, destinationRegister: Register): Set<Match> {
+        return findMatchesWithInstructionMaker(
+            node,
+            valuePatterns,
+        ) { metadata: MatchMetadata, pattern: ValuePattern ->
+            { mapping: ValueSlotMapping ->
+                pattern.makeInstance(
+                    SlotFill(
+                        mapping,
+                        metadata.registerFill,
+                        metadata.constantFill,
+                    ),
+                    destinationRegister,
+                )
+            }
+        }
+    }
+
+    override fun findMatchesForSideEffects(node: CFGNode): Set<Match> {
+        return findMatchesWithInstructionMaker(
+            node,
+            sideEffectPatterns,
+        ) { metadata: MatchMetadata, pattern: SideEffectPattern ->
+            { mapping: ValueSlotMapping ->
+                pattern.makeInstance(
+                    SlotFill(
+                        mapping,
+                        metadata.registerFill,
+                        metadata.constantFill,
+                    ),
+                )
+            }
+        }
+    }
+
+    override fun findMatchesForCondition(node: CFGNode, destinationLabel: BlockLabel, jumpIf: Boolean): Set<Match> {
+        return findMatchesWithInstructionMaker(
+            node,
+            conditionPatterns,
+        ) { metadata: MatchMetadata, pattern: ConditionPattern ->
+            { mapping: ValueSlotMapping ->
+                pattern.makeInstance(
+                    SlotFill(
+                        mapping,
+                        metadata.registerFill,
+                        metadata.constantFill,
+                    ),
+                    destinationLabel,
+                    jumpIf,
+                )
+            }
+        }
+    }
+
+    data class MatchMetadata(
+        val registerFill: MutableMap<RegisterLabel, Register> = mutableMapOf(),
+        val constantFill: MutableMap<ConstantLabel, CFGNode.Constant> = mutableMapOf(),
+        val toFill: MutableMap<ValueLabel, CFGNode> = mutableMapOf(),
+        var size: Int = 0,
+    )
+
+    // Internal function to save some boilerplate from three exported functions.
+    // instructionMaker is truly instructionMakerMaker, as it's some helper to generate instructionMakers.
+    private fun <PatternType : Pattern> findMatchesWithInstructionMaker(
+        node: CFGNode,
+        patterns: List<PatternType>,
+        instructionMaker: (MatchMetadata, PatternType) -> InstructionMaker,
```

Comment:
I like the `instructionMakerMaker` comment :)
However, in all seriousness, how do you feel about renaming this parameter to `createInstructionMaker` or something similar, with a verb in the name? Maybe this would better indicate the purpose of the argument?